first draft of lab2

filipeeduardo opened the pull request.

Commit:
first draft of lab2

powerjg commented:
Quick note (I haven't read this, yet, btw): We need to make it clear that they *cannot* add any more modules to the CPU except for simple things like === and muxes. 

I *will* allow them to change the control unit I/O, if they want.

They should only turn in control.scala and cpu.scala.

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+---
+authors: Jason Lowe-Power
```
You are also an author!

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+
+- [on the RISC-V website](https://riscv.org/specifications/)
+
+|31                 25|24  20|19 15|14     12|11                 7|6       0|
```
Something is broken in this table. Maybe this line? I think it's missing a | at the end.

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+
+Given these inputs, you must generate the correct output on the taken wire.
+The template code from `src/main/scala/components/branch-control.scala` is shown below.
+You will fill in where it says *Your code goes here*.
```
Add something here about how they can use `<`, `>`, `===`, and `=/=`, as well as .toUInt and .toSInt

powerjg commented on lab2/lab2.md:
````diff
@@ -1,30 +1,624 @@
+
+
+```
+dinocpu:sbt> testOnly Lab2 / dinocpu.BranchControlTesterLab2
````
We'll make it so they can just do `testOnly dinocpu...` and they don't need the `Lab2`.

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+
+Feel free to add your own tests in `src/tests/scala`, modify the current tests, and add `print` statements in the tests.
+
+## Draw a diagram for implementing the branch-control unit
```
You mean for implementing branch instructions???

I think we want them to just turn in one diagram which as everything. However, it's going to be pretty messy.

The key things I want to see in the diagram is that they have the correct muxes in the correct places. We should brainstorm this a bit. @jlperona any ideas?

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+
+The following table specifies the `Opcode` format and the control signals to be generated.
+
+|---------|-------------|--------|---------|-------|--------|----------|-----------|----------|---------|------|
```
Also broken.

powerjg commented on lab2/lab2.md:
````diff
@@ -1,30 +1,624 @@
+
+```
+ 
+ branch	:  true if branch or jal and update PC with immediate
````
Try to align these.

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+The following table specifies the `Opcode` format and the control signals to be generated.
+
+|---------|-------------|--------|---------|-------|--------|----------|-----------|----------|---------|------|
+| opcode  |opcode format| branch | memread | toreg |   add  | memwrite | immediate | regwrite | alusrc1 | jump |
```
We don't want to give them this table. This is the main part of the assignment. They should come up with all of these values on their own!

Also, we should make sure to erase this from the history of this repo so they can't see it :). Hopefully they aren't following this discussion.

We can give them a couple of examples (e.g., R-types and invalid), but we shouldn't give them everything.

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+  })
+
+  val signals =
+    ListLookup(io.opcode,
```
We need to explain this syntax in detail so they can use it. Explain how it is just a copy of the table above. Also explain how the opcode is checked against the bit pattern in the first entry and then the signals are in order.

powerjg commented on lab2/lab2.md:
````diff
@@ -1,30 +1,624 @@
+To run the tests, you execute the `SingleCycleTesterLab2` suite as follows:
+
+```
+dinocpu:sbt> testOnly Lab2 / dinocpu.SingleCycleTesterLab2
````
Ditto above.

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+## Checklist
+
+- [ ] You have commented out or removed any extra debug statements.
+- [ ] You have uploaded two files: `cpu.scala` and `alucontrol.scala`.
```
Update.

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+- Start early! There is a steep learning curve for Chisel, so start early and ask questions on Piazza and in discussion.
+- If you need help, come to office hours for the TAs, or post your questions on Piazza.
+
+## Printf debugging
```
Can remove this and below.

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+
+You will upload the two files that you changed to Gradescope on the [Lab 1 - Code](https://www.gradescope.com/courses/35106/assignments/143133) assignment.
+
+- `src/main/scala/components/alucontrol.scala`
```
Update. control.scala, cpu.scala, and branchcontrol.scala

powerjg commented on lab2/lab2.md:
```diff
@@ -1,30 +1,624 @@
+See [the Submission section](#Submission) for more information on how to submit to Gradescope.
+
+|                    |     |
+|--------------------|-----|
```
Needs to be updated :)

powerjg reviewed (commented):
This is a great start! Thanks, Filipe!!!

I take back the thing I said above about changing the I/O on control unit. That doesn't work if they are going to design things.

Maybe, we should actually give them the code in the cpu.scala and just ask to implement the control logic? That's probably too little.

Maybe we should ask them to implement the test??? I'm not sure. I need to think about this more. I'm not sure how useful the control unit test really is, TBH. How should we deal with this... ideas, @jlperona and @filipeeduardo?

powerjg commented:
You know what I'm going to do: I'm going to give them the diagram of everything. However, they have to figure out the control unit and they have to translate the diagram into Chisel. I think this makes the most sense.

powerjg commented:
One more thing we're going to do: We aren't going to give them the control tester. We're going to keep that "secret" on gradescope. They are welcome to create their own, if they want, but they don't *have* to. However, they *do* have to pass the test.

jlperona commented:
> 
> 
> You know what I'm going to do: I'm going to give them the diagram of everything. However, they have to figure out the control unit and they have to translate the diagram into Chisel. I think this makes the most sense.

If you're trying to replicate the Logisim assignments, then this makes sense. They have the diagram in the book that they implement in Logisim; here, you're giving them the diagram (since Chisel is different than Logisim) and they implement in Chisel. And they have to implement the control unit either way. I vote for this idea.

>
> 
> One more thing we're going to do: We aren't going to give them the control tester. We're going to keep that "secret" on gradescope. They are welcome to create their own, if they want, but they don't _have_ to. However, they _do_ have to pass the test.

This is because they could just look at the control tester to figure out the values, no?

We'll probably want to amend this commit soon...

powerjg commented:
@filipeeduardo: feel free to ignore most of my specific comments above. We need to re-work this assignment to be more clear and directed. Here's the outline.

- Hook up the control unit for R-types
  - We will give them the R-type control signals
  - Test: Run r-type tests (use same ones from last time)
- Implement I-type instructions
  - Test: Run I-type instructions
- Implement lw
  - Run lw instructions
- Implement U-type instructions
  - Run auipc/liu
- Implement sw
  - Run sw instruction
- Implement lb, lh, lbu, lhu, sb, sh
  - Run some instruction tests
- Implement branch instructions
  - Implement branch control
    - Test: branch control
  - Hook up branches
    - Test: Run branch instructions
- Implement jal
  - Run jal
- Implement jalr
  - run jalr
- Run full applications.

What I need most from you right now, @filipeeduardo, is to implement all of these tests. I can take care of the writing for the assignment. Thanks for getting it started!

Each test should be implemented as a class that extends `PeekPokeTester`. This way, it can be used both when the students run it an in the grading. Don't worry about the grading-specific tests right now, that can be added after we release the lab.

For the instruction tests, you may need to add some more binaries. You should *definitely* add some full applications. IIRC, Nima pointed out a couple that we could use.

I don't think it makes any sense to add control unit tests. This would essentially just be the answer. Let me know if you disagree.

This need to be done *ASAP*. Let me know what you need from me. I'll be working on my part.

powerjg commented:
@filipeeduardo: I've re-written lab 2. I need you to do the following:

1. Make the tests and document how to run them.
2. Update the lab2.md document with missing details about instructions
3. Double check that I have the correct opcode, etc for all of the instructions.
4. Double check the new uploaded single-cycle diagram to make sure I've gotten all of the wires.

powerjg committed:
Add figure and more details to top.

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

Commit:
added test commands

powerjg committed:
Minor updates to lab2

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

Commit:
Update beginning through part IV of lab2.md

There's a lot to go through still, I need to take a break for a bit.
I'll aim to finish this tonight, though.

Commit:
Finish rest of lab2.md document, remove previous provided files

Final step: one last check through the document

Commit:
Fix broken tables on GitHub's Markdown rendering

Minor changes elsewhere. Should be good to go, apart from adding the
Google form link, and the grading issue that I'll point out as a review.

jlperona commented on lab2/lab2.md:
```diff
@@ -1,30 +1,755 @@
+This time, instead of uploading a paper version to Gradescope, you will give feedback via a Google form.
+Note that the assignment will be out of 90 points and the last 10 points from the feedback will appear in Canvas sometime later.
+
+<link to google form here>
```
Still need to give the link to the Google form.

jlperona commented on lab2/lab2.md:
```diff
@@ -1,30 +1,755 @@
+
+| Name                  | Percentage                 |
+|-----------------------|----------------------------|
+| Each instruction type | 8% each (× 10 parts = 80%) |
```
There's only 9 parts above, since the 10th part is the full programs. Are the branch control unit's tests considered part of the scoring here, and that's where the "10th" part comes from?

powerjg committed:
Minor updates to lab2 based on Justin's feedback
